# HW3

Kevin Wurangian, Fan Hong, Bingqing Li, Yinan Yang, Barrett Poth

In [1]:
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

import pandas as pd
pd.set_option('display.max_colwidth', -1)

import nltk
from nltk.corpus import stopwords

from collections import Counter

from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

#### Task B and C

We did mexican restaurants from Yelp. We scraped 9157 reviews and 98 restaurants.

In [2]:
# read in the reviews into a pandas dataframe
review=pd.read_csv("rest_reviews.csv")

In [3]:
# sanity check
review.head()

restaurants  \
0  Zocalo Café                       
1  Lazarus Brewing                   
2  Iron Cactus                       
3  Las Cazuelas Mexican Restaurant   
4  Uncle Julio’s                     

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      review_contents  
0  I LOVED this place! A great Mexican place!\n\nThe food was amazing and the environment great! Super friendly people working there and a cute restaurant!\n\nWe were in 5 people and we could taste a lot of different meals. Everything was well served, except for my place (ceviche) that was relatively small but still very good! For my taste, a little bit spicy though. \n\nFor sure, the best thing was the queso! I couldn't stop eating the chips with it! They also have a lot of different meals, like tacos, quesadilla, salad, burrito, soup...\n\nThe price was very worth it to pay! If you like a very cool and small place with well made mexican food, I would definitely recommend!                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
1  Happened to make it by the opening day, Christmas Eve, while passing through Austin. Had the French Saisson, which was light floral almost a hint of lemon refreshment. The English IPA was malty with a finish of hops. There are several large group tables, bar seating, lounge area, and outdoor seating. They offer a small food menu mainly consisting of tacos and various snacks.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      

In [4]:
review.shape

(9157, 2)

In [5]:
# Output all text into a single string
text=review["review_contents"].values.tolist()
text=" ".join(text)
text = text.lower()

In [6]:
# tokenize the text 
tokenizer = nltk.tokenize.RegexpTokenizer(r'\w+')
tokens = tokenizer.tokenize(text)

In [7]:
# let's do a word count
words = Counter()

# update counter with new words
words.update(tokens)

In [8]:
stopwords = stopwords.words('english')[:]

In [9]:
filter_words = Counter(x for x in tokens if x not in stopwords and x.isdigit() == False and len(x) != 1 and x.isalpha() == True)

In [10]:
filter_words.most_common()[:50]

[('food', 6612),
 ('tacos', 5372),
 ('good', 5310),
 ('place', 4698),
 ('great', 4087),
 ('taco', 3998),
 ('like', 3035),
 ('one', 2898),
 ('get', 2869),
 ('service', 2848),
 ('austin', 2618),
 ('really', 2526),
 ('back', 2384),
 ('go', 2376),
 ('time', 2300),
 ('salsa', 2112),
 ('would', 2065),
 ('also', 2055),
 ('got', 1957),
 ('delicious', 1929),
 ('chicken', 1798),
 ('ordered', 1791),
 ('best', 1790),
 ('mexican', 1695),
 ('order', 1690),
 ('us', 1651),
 ('chips', 1579),
 ('breakfast', 1579),
 ('came', 1506),
 ('even', 1450),
 ('well', 1448),
 ('restaurant', 1430),
 ('love', 1427),
 ('pretty', 1418),
 ('amazing', 1386),
 ('little', 1385),
 ('definitely', 1382),
 ('try', 1377),
 ('nice', 1372),
 ('menu', 1275),
 ('queso', 1270),
 ('come', 1258),
 ('drinks', 1224),
 ('much', 1158),
 ('friendly', 1153),
 ('made', 1143),
 ('could', 1134),
 ('better', 1127),
 ('first', 1118),
 ('two', 1096)]

i) service (e.g., speed, friendliness, etc.), (ii) food (e.g., quality, taste, etc.), (iii) price and (iv) location (e.g., parking, easy to find, drive through, etc.). 

In [57]:
service_words = ['service', 'quickly', "attentive", 'quick','fast','friendly','helpful', 'waited','wait','rude']
food_words = ['mediocre', 'fried', 'flavorful','quality','authentic','delicious','fresh','tasty','taste','spicy','bland','yummy','yum','variety','tender','juicy','seasoned','greasy','salty','creamy']
price_words =["worth",'prices','cheap','affordable','overpriced','pricey', 'price', 'happy hour']
location_words =['location', 'street','place','parking','downtown', 'sketchy', 'hole in the wall', 'drive through', 'drive thru', 'easy to find']    

In [59]:
s1 = pd.Series(service_words, name = 'service')
s2 = pd.Series(food_words, name = 'food')
s3 = pd.Series(price_words, name = 'price')
s4 = pd.Series(location_words, name = 'location')

replacement_table = pd.concat([s1,s2,s3,s4], axis = 1)

replacement_table = replacement_table.replace(np.nan, '', regex=True)

In [60]:
replacement_table

service       food       price          location
0   service    mediocre   worth       location        
1   quickly    fried      prices      street          
2   attentive  flavorful  cheap       place           
3   quick      quality    affordable  parking         
4   fast       authentic  overpriced  downtown        
5   friendly   delicious  pricey      sketchy         
6   helpful    fresh      price       hole in the wall
7   waited     tasty      happy hour  drive through   
8   wait       taste                  drive thru      
9   rude       spicy                  easy to find    
10             bland                                  
11             yummy                                  
12             yum                                    
13             variety                                
14             tender                                 
15             juicy                                  
16             seasoned                               
17             greasy                                 
18             salty                                  
19             creamy

In [61]:
def replace_names(text):
    
    text = text.lower()
    
    service_words = ['service', 'quickly', "attentive", 'quick','fast','friendly','helpful', 'waited','wait','rude']
    food_words = ['mediocre', 'fried', 'flavorful','quality','authentic','delicious','fresh','tasty','taste','spicy','bland','yummy','yum','variety','tender','juicy','seasoned','greasy','salty','creamy']
    price_words =["worth",'prices','cheap','affordable','overpriced','pricey', 'price', 'happy hour']
    location_words =['location', 'street','place','parking','downtown', 'sketchy', 'hole in the wall', 'drive through', 'drive thru', 'easy to find']
    
    for w in service_words:
        try:
            text = text.replace(w, ' service ')
        except:
            pass
    for w in food_words:
        try:
            text = text.replace(w, ' food ')
        except:
            pass
    for w in price_words:
        try:
            text = text.replace(w, ' price ')
        except:
            pass
    for w in location_words:
        try:
            text = text.replace(w, ' location ')
        except:
            pass
    return text

In [12]:
# apply replace_names() on each review
text_column = []
for r in review.itertuples():
    text = r.review_contents
    new_text = replace_names(text)
    text_column.append(new_text)

In [13]:
# replace old text with new text
review.review_contents = text_column

#### Task D1.

In [14]:
re_len = len(review)

In [15]:
review_attri_count = []
attri = ['service','food','price','location']
for i in range(re_len):
    re_attri_list = []
    for a in attri:
        re_count = review['review_contents'][i].count(a)
        re_attri_list.append(re_count)
    review_attri_count.append(re_attri_list)

In [16]:
attribute = np.array([1,1,1,1])
cos_simi = []
for i in range(re_len):
    cos_matrix = cosine_similarity([attribute, np.array(review_attri_count[i])])
    cos = cos_matrix[0][1]
    cos_simi.append(cos)

In [17]:
review['cos_simi'] = cos_simi
review = review.sort_values(by='cos_simi',ascending=False)
re_high_cos = review.iloc[:200]

In [18]:
re_high_cos.head()

restaurants  \
7975  El Arroyo 5th Street              
5731  Las Cazuelas Mexican Restaurant   
6590  Angie’s Mexican Restaurant        
8057  Taco Joint                        
8425  El Chile Café Y Cantina           

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                review_contents  \
7975  let's see...what can i say about this  location ...oh, that's right, it's awesome. \n\nnot just for their hilarious signs that are shared globally on facebook, but their drinks, food, and  service  are phenomenal.\n\ndo you like  price ? me too. know who else does? el arroyo. seriously, you should check it out. \n\nthey have a great patio as well. do yourself a favor and get one (or a few) of their ritas, some chips & queso, and soak in all the amazing weather austin has to offer.\n\nby the way, they have live music on the weekends. sounds like heaven right?                                                                                                                                                                                                                                                                                        
5731  good break service  tacos!  the migas (soco taco) was good.  the chorizo and egg was also good; a bit  food  -- but better than dry.  the pollo flaco was just ok; would have rather had avocado than their "guacamole."  overall, the convenience of ordering and paying online and just having to pop in to grab the bag makes this  location    price   visiting.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
6590  taco shack is by no means one of the better taco joints in town, but the shack taco and the el niño taco hit the spot every time, the  service  is practically instantaneous, and the  price  +  location  is right on. \n\nthere's plenty of really negative reviews on here, but i guarantee you those negative nancys all eat here at least twice a month.  you know how whataburger always seems exquisitely  food  when you pick it up on the way home from a bar late-late saturday night?  taco shack is the exact same way, except at 7:30 am on a weekday.  and you're sober.                                                                                                                                                                                                                                                                                      
8057  i feel ashamed of not knowing this  location  earlier, but i'm glad i did now. the tortillas are both crispy on the outside and fluffy on the inside. it seems like all the ingredients are very  food . the  price  is such a bargain and there is an assortment of sauces you can put on the tacos\n\nsuch as habanero, peanut, avocado, chipotle, green, and red.\n\ni loved a

#### Task D2.

In [19]:
analyser = SentimentIntensityAnalyzer()

In [20]:
review_sentiment = []
for r in re_high_cos.itertuples():
    snt = analyser.polarity_scores(r.review_contents)
    review_sentiment.append(snt)
review_sentiment = [r['compound'] for r in review_sentiment]

In [39]:
re_high_sent = re_high_cos
re_high_sent['sentiment'] = review_sentiment
re_high_sent = re_high_cos.sort_values(by='sentiment',ascending=False)

In [22]:
re_high_sent.head()

restaurants  \
5732  Kesos Tacos                            
5616  Las Cazuelas Mexican Restaurant        
4772  Z’Tejas Mexican Restaurant and Grill   
5071  Suerte                                 
1182  Licha’s Cantina                        

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           review_contents  \
5732  ok, to be honest, i kind of don't want to give this review. i want to keep this  location  a secret... all to myself. but i see the greater good and i realize that the more biz these guys get, the longer they'll be around. so here ya go and here it is.\n\nthis  location  is one of my all time go-to favorites. i grew up in san antonio, by the way, and i pretty much love everything here. the break service  tacos are are incredible and they now offer them in ultra-huge sizes. the caldo is fantastic and the plates are  food  and scrumptious. they have horchata and mexican sodas and their salsas (yes, they have like 7 different ones) are insanely  food  and good (and hot!!!). they have menudo!\n\nmy recommendation is the large quesadilla, however. they make 'em huge and wonderful and are surprisingly sprinkled with a parmesan-like cheese on top. \n\ni also love to get a couple of beers here every now and again. they have my brand (lone star) and it's   price  . they are also open late and i often hit them up after the bars close for a taco or two. \n\nyou may have to brush up on your spanish, though. sometimes i have to speak entirely in spanish to order. i like it though. it's fun to keep the old language skills sharp, yeah?                                                                                                                                                                                                                                                                                                               
5616  top notch  location . food a++  service  a++ ambiance a++\n\nthe food is not your typical mexican that you can get everywhere, it has very unique and  food  interior mx food. i was blown away by the  food  and the flavors. while i did try a boring carnitas dish, my friends all had unique items that i was overly impressed with. i knew i'd be drinking later so i didn't want to be too adventurous, so i definitely will need to go back to enjoy the other menu items.\n\nthe portions are huge, so it wasn't the best choice before heading  location , as i would have loved to take half home to enjoy later. you could definitely share if you'd like. th

In [23]:
# create dictionaries: key is rest name, value is list of sentiments and cos similarities for that rest
rest_cs_dict = {}
rest_sent_dict = {}
for r in re_high_cos.itertuples():
    
    if r.restaurants in rest_sent_dict:
        rest_sent_dict[r.restaurants].append(r.sentiment)
    else:
        rest_sent_dict[r.restaurants] = [r.sentiment]
        
    if r.restaurants in rest_cs_dict:
        rest_cs_dict[r.restaurants].append(r.cos_simi)
    else:
        rest_cs_dict[r.restaurants] = [r.cos_simi]        

In [24]:
# get average of list
def avg(lst): 
    return sum(lst) / len(lst) 

In [25]:
# create dictionaries: key is rest name, value is avg sentiment and cos similarities for that rest
rest_cs_avg_dict = {}
rest_sent_avg_dict = {}
for r in rest_cs_dict:
    rest_cs_avg_dict[r] = avg(rest_cs_dict[r])
    rest_sent_avg_dict[r] = avg(rest_sent_dict[r])

In [26]:
# get sorted lists of rest_cs_avg_dict and rest_cs_avg_dict
import operator
rest_cs_avg = sorted(rest_cs_avg_dict.items(), key=operator.itemgetter(1))
rest_cs_avg.reverse()

rest_sent_avg = sorted(rest_sent_avg_dict.items(), key=operator.itemgetter(1))
rest_sent_avg.reverse()

We realized there is little variability in the restaurants average cosine similarities, while there is higher variability in the restaurants sentiment scores. Knowing this, we decided to multiply the two values together, cosine similarity and sentiment scores, to find our top recommendations. We knew that the cosine similarities would not make too big of a difference in the sentiment scores which is why we felt comfortable doing this.

In [27]:
# create dictionary: key is rest name, value is sentiment score * cosine similarity for that restaurant
rest_cs_sent_dict = {}
for r in rest_cs_dict:
    cs = rest_cs_avg_dict[r]
    sent = rest_sent_avg_dict[r]
    rest_cs_sent_dict[r] = cs * sent

In [28]:
# get sorted list of rest_cs_send_dict
rest_cs_sent = sorted(rest_cs_sent_dict.items(), key=operator.itemgetter(1))
rest_cs_sent.reverse()

#### Recommendations

In [29]:
recommendations = [r[0] for r in rest_cs_sent[:3]]
for i in range (len(recommendations)):
    output = str(i+1) + '. Name: ' + recommendations[i]
    output += '\n    Cos: ' + str(rest_cs_avg_dict[recommendations[i]])
    output += '\n   Sent: ' + str(rest_sent_avg_dict[recommendations[i]])
    print (output, end='\n\n')

1. Name: Gloria’s Latin Cuisine
    Cos: 1.0
   Sent: 0.9668

2. Name: Taqueria Los Altos
    Cos: 0.974341649025257
   Sent: 0.9911000000000001

3. Name: Taco Ranch
    Cos: 0.9707253433941511
   Sent: 0.9932



### Task F

In [30]:
ratings=pd.read_csv("rest_ratings.csv")

In [31]:
r5='alt="5.0 star rating'
r4='alt="4.0 star rating'
r3='alt="3.0 star rating'
r2='alt="2.0 star rating'
r1='alt="1.0 star rating'

In [32]:
new_rating=[]
for n in ratings.itertuples():
    if r5 in n[2]:
        new_rating.append(5)
    elif r4 in n[2]:
        new_rating.append(4)
    elif r3 in n[2]:
        new_rating.append(3)
    elif r2 in n[2]:
        new_rating.append(2)
    elif r1 in n[2]:
        new_rating.append(1)
    else:
        new_rating.append(0)
    

In [33]:
ratings['ratings']=new_rating

In [34]:
len(new_rating)

9157

In [35]:
#All of the restaurants have their own ratings.
for n in ratings.itertuples():
    if n[2]==0:
        print('yes')

In [36]:
name_list=[]
average_list=[]
for m in ratings['restaurants'].drop_duplicates():
    index=0
    count=0
    for n in ratings.itertuples():
        if m==n[1]:
            index+=1
            count=count+n[2]
    average_list.append(count/index)
    name_list.append(m)

In [37]:
b=dict(zip(name_list,average_list))
rank=sorted(b.items(),key = lambda x : x[1],reverse=True)

In [42]:
rank[0:10]

[('Taquizas Mariana', 5.0),
 ('Las Abuelas', 5.0),
 ('Taqueria Lucerito', 4.5),
 ('Antojitos Guatemala', 4.4),
 ('Little Plaza Taco Truck', 4.375),
 ('Veracruz All Natural', 4.111111111111111),
 ('Rules & Regs', 4.111111111111111),
 ('Baja St Tacos & Coastal Cuisine', 4.09375),
 ('Arturo’s Underground Cafe', 4.0828402366863905),
 ('El Alma', 4.063157894736842)]

In [56]:
rest_cs_avg_dict['Veracruz All Natural']

0.9677634565778661

In [54]:
rest_sent_avg_dict['Veracruz All Natural']

0.71354

In [50]:
rest_cs_avg_dict['Arturo’s Underground Cafe']

0.9609328918804079

In [53]:
rest_sent_avg_dict['Arturo’s Underground Cafe']

0.8824

In [49]:
rest_cs_avg_dict['El Alma']

0.9696534301640931

In [52]:
rest_sent_avg_dict['El Alma']

0.541525

These recommendations do not align with our recommendations. These restaraunts did not even make it into our top 200 cosine similarity list. This means that these restaurants would not satisfy the requirements of the user looking for the recommendations. We also do not know the sentiments of these restaurants for the same reason, they did not even make it in our top 200 cosine similarities list.

Once realizing this, we found the top three out of the recommendations from task f that did make it into our top 200 cosine similarity list. Of these, the Veracruz All Natural, Arturo’s Underground Cafe, and El Alma, they did have very high cosine similarity. All of their sentiments were positive as well with scores of .71, .88, and .54 respectively. From the rating suggestions, these restaurants were ranked 6, 9, and 10 respectively. These restaurants would meet the requirements of the user looking for recommendations.